sound/soc/intel/boards/sof_sdw.c

```diff
@@ -411,25 +411,30 @@ static int create_codec_dai_name(struct device *dev,
 
 static int set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
 			       struct snd_soc_dai_link *dai_links,
-			       bool playback)
+			       bool playback, int group_id)
 {
 	int i;
 
-	for (i = 0; i < link->num_adr; i++) {
-		unsigned int part_id;
-		int codec_index;
-
-		part_id = SDW_PART_ID(link->adr_d[i].adr);
-		codec_index = find_codec_info_part(part_id);
+	do {
+		for (i = 0; i < link->num_adr; i++) {
+			unsigned int part_id;
+			int codec_index;
 
-		if (codec_index < 0)
-			return codec_index;
+			part_id = SDW_PART_ID(link->adr_d[i].adr);
+			codec_index = find_codec_info_part(part_id);
 
-		if (codec_info_list[codec_index].init)
-			codec_info_list[codec_index].init(link, dai_links,
-						 &codec_info_list[codec_index],
-						 playback);
-	}
+			if (codec_index < 0)
+				return codec_index;
+			if (link->adr_d[i].endpoints->group_id != group_id)
+				continue;
+			if (codec_info_list[codec_index].init)
+				codec_info_list[codec_index].init(link,
+						dai_links,
+						&codec_info_list[codec_index],
+						playback);
+		}
+		link++;
+	} while (link->mask && group_id);
```

Comment:
@bardliao I get the logic by looking at the conversation about you had with @ranj063 but based on just the git commit message and code, this check for "while(group_id)" still seems cryptic. Maybe add a comment stating that group_id>0 means codec is aggregated

Reply:
Done

Reply:
@bardliao a clearer comment would be something that says something like: "The loop initializes the codec if it does not belong to an aggregated group or initializes the first codec in an aggregated group and iterates through the links to initialize the rest of the codecs in the same group"

Reply:
Agree, let's improve the comment a bit. In fact this whole aggregation implementation seems a bit inconvenient to handle. I understand that it has been added as an after-thought. So what we have now is the outer loop over links in `sof_card_dai_links_create()`, which calls `create_sdw_dailink()` which then continues looping over the same list of links, looking for groups. Then `set_codec_init_func()` is called which now again loops over the same links... It isn't anything exceptional, but maybe someone comes up with a good idea to simplify this a bit.

Reply:
Thanks @ranj063 @lyakh The comment is updated with slight modification from @ranj063 's suggestion. A link may contain more than one codec and we will go through all of them in the first iteration. An assumption here is all codecs in the same link will either not aggregated or with in the same aggregated group. Note that the adr link here doesn't mean the physical link. A physical link can be separate to different adr links.

Reply:
Thanks @bardliao -- I'm afraid the comment is not helping enough yet. Even with @ranj063 's changes, there's still no explanation of the special meaning of group_id==0. How about:

"Initialize the codec. If codec is part of an aggregated group (group_id>0), initialize all codecs belonging to same group."

?

Reply:
> Thanks @bardliao -- I'm afraid the comment is not helping enough yet. Even with @ranj063 's changes, there's still no explanation of the special meaning of group_id==0. How about:
> 
> "Initialize the codec. If codec is part of an aggregated group (group_id>0), initialize all codecs belonging to same group."
> 
> ?

Thanks @kv2019i That is clear and simple.